src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+(define-constant ERR_NAMESPACE_PREORDER_LAUNCHABILITY_EXPIRED 1010)
+(define-constant ERR_NAMESPACE_OPERATION_UNAUTHORIZED 1011)
+(define-constant ERR_NAMESPACE_STX_BURNT_INSUFFICIENT 1012)
+(define-constant ERR_NAMESPACE_BLANK 1013)
+(define-constant ERR_NAMESPACE_ALREADY_LAUNCHED 1014)
+(define-constant ERR_NAMESPACE_HASH_MALFORMED 1015)
+(define-constant ERR_NAMESPACE_CHARSET_INVALID 1016)
+
+(define-constant ERR_NAME_PREORDER_NOT_FOUND 2001)
+(define-constant ERR_NAME_PREORDER_EXPIRED 2002)
+(define-constant ERR_NAME_PREORDER_FUNDS_INSUFFICIENT 2003)
+(define-constant ERR_NAME_UNAVAILABLE 2004)
+(define-constant ERR_NAME_OPERATION_UNAUTHORIZED 2006)
+(define-constant ERR_NAME_STX_BURNT_INSUFFICIENT 2007)
+(define-constant ERR_NAME_EXPIRED 2008)
+(define-constant ERR_NAME_GRACE_PERIOD 2009)
+(define-constant ERR_NAME_BLANK 2010)
+(define-constant ERR_NAME_ALREADY_CLAIMED 2011)
+(define-constant ERR_NAME_CLAIMABILITY_EXPIRED 2012)
+(define-constant ERR_NAME_NOT_FOUND 2013)
+(define-constant ERR_NAME_REVOKED 2014)
+(define-constant ERR_NAME_TRANSFER_FAILED 2015)
+(define-constant ERR_NAME_PREORDER_ALREADY_EXISTS 2016)
+(define-constant ERR_NAME_HASH_MALFORMED 2017)
+(define-constant ERR_NAME_PREORDERED_BEFORE_NAMESPACE_LAUNCH 2018)
+(define-constant ERR_NAME_NOT_RESOLVABLE 2019)
+(define-constant ERR_NAME_COULD_NOT_BE_MINTED 2020)
+(define-constant ERR_NAME_COULD_NOT_BE_TRANSFERED 2021)
+(define-constant ERR_NAME_CHARSET_INVALID 2022)
+
+(define-constant ERR_PRINCIPAL_ALREADY_ASSOCIATED 3001)
+(define-constant ERR_INSUFFICIENT_FUNDS 4001)
+
+;;;; Constants
+(define-constant BURN_ADDRESS 'S0000000000000000000002AA028H)
+
+;; TTL
+;; todo(ludo): add real-life values
+(define-constant NAMESPACE_PREORDER_CLAIMABILITY_TTL u10)
+(define-constant NAMESPACE_LAUNCHABILITY_TTL u10)
+(define-constant NAME_PREORDER_CLAIMABILITY_TTL u10)
+(define-constant NAME_GRACE_PERIOD_DURATION u5)
+
+(define-constant ATTACHMENTS_INV_PAGE_SIZE u8)
+(define-data-var attachments-inv-index-cursor uint u0)
+(define-data-var attachments-inv-page-cursor uint u0)
+(define-map attachments-inv 
+    ((page uint) (index uint)) 
+    ((content-hash (buff 20))))
+
+;; Price tables
+(define-constant NAMESPACE_PRICE_TIERS (list
+  u96000 
+  u9600 u9600 
+  u960 u960 u960 u960 
+  u96 u96 u96 u96 u96 u96 u96 u96 u96 u96 u96 u96 u96))
+
+;;;; Data
+(define-map namespaces
+  ((namespace (buff 19)))
+  ((namespace-import principal)
+   (revealed-at uint)
+   (launched-at (optional uint))
+   (namespace-version uint)
+   (renewal-rule uint)
+   (price-function (tuple 
+    (buckets (list 16 uint)) 
+    (base uint) 
+    (coeff uint) 
+    (nonalpha-discount uint) 
+    (no-vowel-discount uint)))))
+
+(define-map namespace-preorders
+  ((hashed-salted-namespace (buff 20)) (buyer principal))
+  ((created-at uint) (claimed bool) (stx-burned uint)))
+
+(define-non-fungible-token names (tuple (name (buff 16)) (namespace (buff 19))))
```

Comment:
1) Why `buff` and not `string-ascii` or even a `string-username` sub type? A string seems a better representation for name and namespace data.

2) As a smart contract, I would like e.g. to restrict certain functions to the owner of a username. BNS should have a function like 
```
(define-read-only (get-owner? (username (tuple (name (buff 16)) (namespace (buff 19)))))
  (nft-get-owner? names username)
)
```

Reply:
1. I agree, using ascii-string would have probably been a bit better, however we need to perform so hash operations onchain, and those functions are not compatible with strings at this point. hence buffer types.

2. This contract is exposing the method 
```
(define-read-only (name-resolve (namespace (buff 19)) (name (buff 16)))
```
that does what you'd expect.

Reply:
@lgalabru If I read the code correctly, it return only the zonefile hash `(ok (get zonefile-hash name-props)))`

Reply:
@lgalabru Yes, please return the owner as well :-)

Reply:
Adressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/b92368b9cc0a6ecd013e7b26e173a6b853be50dc